File: src/main/java/org/cognoseed/tweetlazo/TweetDispatcher.java
package org.cognoseed.tweetlazo;

import akka.actor.ActorRef;
import akka.actor.Props;
import akka.actor.UntypedActor;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import akka.japi.Creator;
import twitter4j.HashtagEntity;
import twitter4j.Status;

import java.util.*;

/**
 * Top of the tree of actors, this actor receives tweets from {@link TweetListener} and distributes the tweets to the
 * correct children actors.
 */
public class TweetDispatcher extends UntypedActor {

    private final LoggingAdapter logger = Logging.getLogger(context().system(), this);

    public static Props props(final String[] args) {
        return Props.create(new Creator<TweetDispatcher>() {
            public TweetDispatcher create() throws Exception {
                return new TweetDispatcher(args);
            }
        });
    }

    String[] hashtags;
    Map<String, ActorRef> children = new HashMap<>();

    public TweetDispatcher(String[] hashtags) {
        this.hashtags = hashtags;
    }

    @Override
    public void preStart() throws Exception {
        logger.debug("Creating children for each hashtag...");
        for (String hashtag : hashtags) {
            String key = hashtag.substring(1).toLowerCase();
            children.put(key, context().actorOf(Props.create(GoalCounter.class, key), key));
        }
    }

    @Override
    public void onReceive(Object message) throws Exception {
        if (message instanceof Status) {
            final Status status = (Status) message;
            logger.debug("{} : {}", status.getId(), status.getText());

            // find the hashtags in the tweet that we're tracking, and forward the tweet to them. ONCE.
            Arrays.stream(status.getHashtagEntities())
                    .map(HashtagEntity::getText)
                    .map(String::toLowerCase)
                    .distinct()
                    .filter(children::containsKey)
                    .map(children::get)
                    .forEach(ref -> ref.tell(status, self()));
        } else {
            unhandled(message);
        }
    }
}

File: src/main/scala/org/cognoseed/tweetlazo/TweetDispatcher.scala
package org.cognoseed.tweetlazo

import akka.actor.{Actor, ActorRef, ActorRefFactory, Props}
import akka.event.Logging
import twitter4j.Status

object TweetDispatcher {
  case class WatchHashtag(hashtag: String)
  case class UnwatchHashtag(hashtag: String)

  def props(): Props =
    Props(new TweetDispatcher((factory, key) => factory.actorOf(GoalCounter.props(key), key)))

  def props(maker: (ActorRefFactory, String) => ActorRef): Props = Props(new TweetDispatcher(maker))
}

/** An actor that receives tweets from [[org.cognoseed.tweetlazo.TweetListener]] and distributes the tweets to the
  * correct children actors.
  */
class TweetDispatcher(maker: (ActorRefFactory, String) => ActorRef) extends Actor {
  import GoalCounter._
  import TweetDispatcher._

  val log = Logging(context.system, this)
  var children = Map.empty[String, ActorRef]

  def receive = {
    case tweet: Status =>
      tweet.getHashtagEntities.map(_.getText.toLowerCase).distinct.map(children.get).foreach(_ foreach(_ forward tweet))
    case WatchHashtag(hashtag) =>
      val key = hashtag.toLowerCase
      children = children.updated(key, maker(context, key))
    case UnwatchHashtag(hashtag) =>
      val key = hashtag.toLowerCase
      children.get(key).foreach(context.stop)
      children = children - key
    case msg: Counts => children.get(msg.hashtag.toLowerCase).foreach(_ forward msg)
    case msg: CountsSinceLast => children.get(msg.hashtag.toLowerCase).foreach(_ forward msg)
  }

}
